keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1009 @@
+- Performance considerations must be taken into account, especially for large images or artifacts.
+
+### Vocabulary: OCI Images, Artifacts, and Objects
+
+**OCI Image ([spec](https://github.com/opencontainers/image-spec/blob/main/spec.md)):**
+   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
+     It includes a filesystem bundle and metadata required to run a container.
+   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
+     (which provides configuration data such as environment variables, entry points, etc.).
+   - **Use Case:** Used primarily for packaging and distributing containerized applications.
+
+**OCI Artifact ([guidance](https://github.com/opencontainers/image-spec/blob/main/artifacts-guidance.md)):**
+   - An artifact describes any content that is stored and distributed using the OCI image format.
+     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
+   - Artifacts use the same image manifest and layer structure but may contain different types of data
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps to determining the file location on the host filesystem.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
```

Comment:
I'm coming in later and trying to catch up, so apologies if this is well covered.  But it's not clear from the kep how we are separating out the source (and what form it is in: filesystem, multiple files, or single file) and how a user disambiguates it.

* Conceptually we are already mounting images as filesystems for containers.  
* Volume sources can be both filesystems (simplest form is a directory) or a single file (e.g. secret mount)
* Artifacts may be filesystems, multiple-files with names, single file with a name, or potentially a single file without a name?

Therefore we are talking about the following actions

1. Mount image as filesystem (no qualification beyond image)
2. Mount image as file (somehow point to a file in the image via a path spec)
3. Mount image as directory (somehow point to a directory in the image via a path spec)
4. Mount artifact as filesystem (the artifact must have a "name" that can be represented in a directory)
5. Mount artifact as file (artifact is a singular file, otherwise the user has to disambiguate which artifact?)
6. Mount artifact as directory (artifact either has multiple entries that have names, or is a compressed filesystem, or has a single entry that has a name)

?  What isn't clear to me is how the user specifies 2 and 3, and then whether artifacts are guaranteed to have names (which determines what has to be provided by the user for 5 and 6).

Reply:
Imagine an OCI artefact that is an unnamed raster image. Eventually, the thing we design here might want to try to mount that.

However, I think we have an option to intentionally make that example case a non-goal for this KEP (and instead suggest using an alternative, such as a CSI driver). Any CSI driver needs to know how to turn the raster data into a volume, even if it's as simple as “take the raw binary data and make it available as a file called _x_”.

There are two obvious places to put a boundary (at least, that's how I see it):
- we'd try to handle all things that can be fetched from an OCI registry, and the boundary is whether we want to just mount it into a Pod or run it as a container
- we'd (eventually?) like to handle lots of things that can be fetched from an OCI registry, and the boundary is whether we want the CRI-level container runtime to do the fetching or not. Container runtimes only know about fetching container-like images / we can't assume they have additional capabilities.

---

Overall, I'd like to do the simplest thing we can live with and not find ourselves painted into an API corner. Taking the single-file case out of the KEP looks good, especially if it gets us to a point where we can merge it as provisional.

Reply:
> Artifacts may be filesystems, multiple-files with names, single file with a name, or potentially a single file without a name?

At the lowest level artifacts are basically "ship things that aren't tarballs wrapped with the same JSON as OCI container images".

https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage

The WASM OCI artifact type docs are a good example: https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/

The use case of "mount AI model weights stored as an OCI artifact" to me seems most useful...it's generic *architecture independent* content so packing it as a container introduces logistical overhead (need to "build" it for multi-arch scenarios even though it isn't).

I actually don't know if there's an existing spec for it...would be useful if there was!

> Container runtimes only know about fetching container-like images / we can't assume they have additional capabilities.

At least IMO if something like this ends up in core kubelet it'd also get exposed by `podman play kube` etc.

Reply:
I added a section "Filesystem Representation" covering the above point. Please continue the discussion there if there is anything left open or I missed.